### PR TITLE
feat(useTable): support default states and onChange callbacks

### PR DESCRIPTION
**Pull request by ling1726**
After some deliberation, there should be no 'autocontrol' support for features
of the `useTable` hook. The hook is intended to be used standalone so
simple providing default state values on mount and change callbacks
should be enough to allow it to integrate with apps' own state
management.

```ts
const [selectedRows, setSelectedRows] = React.useState([1, 2])

const {
    rows,
  } = useTable({
    columns,
    items,
    selection: {
      onRowSelectionChange: setSelectedRows,
      defaultSelectedRows: selectedRows,
    },
});
```

Addresses #24226

### CODE REVIEW

File: packages/react-components/react-table/src/hooks/useSelection.ts [hunk 1]
```diff
@@ -1,38 +1,70 @@
...
+export function useSelection<TItem>(options: UseSelectionOptions<TItem>): SelectionStateInternal {
+  const { selectionMode, items, getRowId, onRowSelectionChange = () => undefined, defaultSelectedRows } = options;
 
-export function useSelection<TItem>(
-  selectionMode: SelectionMode,
-  items: TItem[],
-  getRowId: GetRowIdInternal<TItem>,
-): SelectionStateInternal {
   const prevSelectionMode = usePrevious(selectionMode);
-  const [selected, setSelected] = React.useState(() => new Set<RowId>());
```
Comment (ling1726) [inline packages/react-components/react-table/src/hooks/useSelection.ts:12]: removed the react state here, there's actually no point in duplicating the same information that is in the selection manager. Instead added a force updater to update every time the selection is updated

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                             |          Size (minified/GZIP) |
| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `189.042 kB`<br />`51.964 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>                            |  `32.895 kB`<br />`10.778 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/ee77a76375e07c52729fdac1a08471607c12652d'>ee77a76375e07c52729fdac1a08471607c12652d</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/Avatar-mount_base.html">1284</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/Avatar-mount.html">1263</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/Button-mount_base.html">965</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/Button-mount.html">947</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/FluentProvider-mount_base.html">1596</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/FluentProvider-mount.html">1577</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/FluentProviderWithTheme-mount_base.html">545</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/FluentProviderWithTheme-mount.html">625</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">515</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">596</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">633</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">629</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/MakeStyles-mount_base.html">1881</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/MakeStyles-mount.html">1930</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/SpinButton-mount_base.html">2508</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24403/perf-test-react-components/SpinButton-mount.html">2490</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: packages/react-components/react-table/src/hooks/useSort.ts [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
+  columns: ColumnDefinition<TItem>[],
+  onSortColumnChange: UseTableSortOptions['onSortColumnChange'] = () => undefined,
+  defaultSortColumn?: UseTableSortOptions['defaultSortColumn'],
+): SortStateInternal<TItem> {
+  const [sorted, setSorted] = React.useState<SortColumnState>(
+    defaultSortColumn ?? {
+      sortDirection: 'ascending' as const,
+      sortColumn: undefined,
+    },
+  );
```
Comment (bsunderhus) [inline packages/react-components/react-table/src/hooks/useSort.ts:14]:
```suggestion

const defaultSortColumnState: SortColumnState =  {
      sortDirection: 'ascending' as const,
      sortColumn: undefined,
    }

export function useSort<TItem>(
  columns: ColumnDefinition<TItem>[],
  onSortColumnChange: UseTableSortOptions['onSortColumnChange'] = () => undefined,
  defaultSortColumn?: UseTableSortOptions['defaultSortColumn'],
): SortStateInternal<TItem> {
  const [sorted, setSorted] = React.useState<SortColumnState>(defaultSortColumn ?? defaultSortColumnState);
```

**Review by bsunderhus — APPROVED** [review]

**Comment by layershifter** [comment]
```
    selection: {
      onRowSelectionChange: setSelectedRows,
      defaultSelectedRows: selectedRows,
    },
```

I did not get this part. If not's autocontrolled why there is `default` prefix?

**Comment by ling1726** [comment]
> I did not get this part. If not's autocontrolled why there is `default` prefix?

By autocontrolled I mean rather, 'it can't be both controlled and uncontrolled' which is what our components normally do. at least that's what I understood by 'autocontrolled' sorry for the confusion.

@layershifter  Does saying that the hook is 'uncontrolled' make more sense ?

**Comment by ling1726** [comment]
as discussed with @layershifter the minbar should be **controlled** state. This is because of how React works. We **want** users to control things through props and not through imperative handles. By removing controlled support it's no longer possible to control `useTable` from a higher component in the hierarchy.

Imagine a scenario where `useTable` manages a data table, and on the top level app bar there is a `clear selection` button. In order for that button to work, then `useTable` would need to bubble up its imperative methods instead of hooking into the app's existing state management

**Pull request closed without merging**